### PR TITLE
bt-19: + implement sign in on server

### PR DESCRIPTION
**Pull request by ptarho**
This is just a first general implementation, because jwt token and verifyLoginCredentials should be done when I will get code from other tasks. Also I have a bunch of questions like where to throw an errors, do I need to replace error messages with corresponding enums or we can leave it for now, ect
  
email_validation
email_wrong
correct
password_wrong

### CODE REVIEW

File: backend/src/bundles/auth/auth.controller.ts [hunk 1]
```diff
@@ -76,6 +93,51 @@ class AuthController extends ControllerBase {
             payload: await this.authService.signUp(options.body),
         };
     }
+    /**
+     * @swagger
+     * /auth/sign-in:
+     *    post:
```
Comment (MaksymGavatuk) [inline backend/src/bundles/auth/auth.controller.ts:101]: Did you check swagger api for endpoint ? That is, did you try to make request using swagger ?

File: shared/src/bundles/users/types/user-sign-in-response-dto.type.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+type UserSignInResponseDto = {
+    id: number;
+    email: string;
+    token: string;
+};
```
Comment (MaksymGavatuk) [inline shared/src/bundles/users/types/user-sign-in-response-dto.type.ts]: Leave please only token as I write in comment for jwt auth PR

File: backend/src/bundles/auth/auth.service.ts [hunk 3]
```diff
@@ -1,21 +1,60 @@
...
+        const user = await this.userRepository.findByEmail(email);
+
+        if (!user) {
+            throw new Error('Incorrect email');
+        }
+
+        // const isEqualPassword = true await cryptCompare(password, password);
+        // if (!isEqualPassword) {
+        //     throw new Error('Password not match');
+        // }
```
Comment (MaksymGavatuk) [inline backend/src/bundles/auth/auth.service.ts]: This is TODO flow, or you forgot to delete this lines ?

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: backend/src/bundles/auth/auth.service.ts (hunk 3 above)
Comment (Monyster) [reply backend/src/bundles/auth/auth.service.ts]: I think it's a TODO because I'm implement cryptCompare in my task #86

File: backend/src/bundles/auth/auth.controller.ts [hunk 4]
```diff
@@ -33,6 +36,20 @@ class AuthController extends ControllerBase {
                     }>,
                 ),
         });
+
+        this.addRoute({
+            path: AuthApiPath.SIGN_IN,
+            method: 'POST',
+            validation: {
+                body: userSignUpValidationSchema,
```
Comment (JosuerBague) [inline backend/src/bundles/auth/auth.controller.ts]: perhaps this should be userSignInValidationSchema?

File: backend/src/bundles/auth/auth.controller.ts [hunk 5]
```diff
@@ -33,6 +36,20 @@ class AuthController extends ControllerBase {
...
+            },
+            handler: (options) =>
+                this.signIn(
+                    options as ApiHandlerOptions<{
+                        body: UserSignUpRequestDto;
```
Comment (JosuerBague) [inline backend/src/bundles/auth/auth.controller.ts]: might be UserSignInResponseDto

File: backend/src/bundles/auth/auth.ts [hunk 6]
```diff
@@ -1,10 +1,10 @@
-import { userService } from '~/bundles/users/users.js';
+import { userRepository, userService } from '~/bundles/users/users.js';
 import { logger } from '~/common/logger/logger.js';
 
 import { AuthController } from './auth.controller.js';
 import { AuthService } from './auth.service.js';
 
-const authService = new AuthService(userService);
+const authService = new AuthService(userService, userRepository);
```
Comment (JosuerBague) [inline backend/src/bundles/auth/auth.ts]: I am not sure if we need to give authService direct access to userRepository since useRepository is accessible via userService? I am not sure though.

File: backend/src/bundles/auth/auth.service.ts (hunk 3 above)
Comment (ptarho) [reply backend/src/bundles/auth/auth.service.ts]: I left this code to show that this part is not finished, becuse I need to use cryptCompare or some analogue that was not implemented at that time. Because currently it`s just check if password='HASH', as it was hardcoded inside sign-up route

File: backend/src/bundles/auth/auth.controller.ts (hunk 1 above)
Comment (ptarho) [reply backend/src/bundles/auth/auth.controller.ts:101]: Yeah request are working in swagger ui

File: backend/src/bundles/auth/auth.ts (hunk 6 above)
Comment (anton-otroshchenko) [reply backend/src/bundles/auth/auth.ts]: > I am not sure if we need to give authService direct access to userRepository since useRepository is accessible via userService? I am not sure though.

I think you are right. You can create a method in userService which accesses userRepository since you alredy using userService in this class

Comment (ptarho) [reply backend/src/bundles/auth/auth.ts]: The main reason I imported repository is because I wanted to make findByEmail method inside user-service return just user object if we will need to use it, while inside auth-service we need userEntity with password hash

Comment (ptarho) [reply backend/src/bundles/auth/auth.ts]: Maybe in this case we can delete user-service and use only user-repository?

File: backend/src/bundles/users/user.service.ts [hunk 7]
```diff
@@ -19,6 +20,30 @@ class UserService implements Service {
         return Promise.resolve(null);
     }
 
+    public async findByEmail(
+        email: UserEntity['email'],
+    ): Promise<UserGetOneItemResponseDto> {
+        const user = await this.userRepository.findByEmail(email);
```
Comment (Monyster) [inline backend/src/bundles/users/user.service.ts]: I guess this is similar to https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/88#discussion_r1305874025
```suggestion
        const user = await this.userRepository.find({email});
```

File: backend/src/bundles/users/user.service.ts [hunk 8]
```diff
@@ -19,6 +20,30 @@ class UserService implements Service {
...
+
+        if (!user) {
+            throw new Error('Incorrect email');
```
Comment (Monyster) [inline backend/src/bundles/users/user.service.ts]: Change new Error to any exception or create new in folder /shared/src/framework/exceptions

File: backend/src/bundles/users/user.service.ts [hunk 9]
```diff
@@ -19,6 +20,30 @@ class UserService implements Service {
...
+        }
+
+        return user.toObject();
+    }
+
+    public async findById(
+        id: UserEntity['id'],
+    ): Promise<UserGetOneItemResponseDto> {
+        const user = await this.userRepository.findById(id);
```
Comment (Monyster) [inline backend/src/bundles/users/user.service.ts]: I guess this is similar to https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/88#discussion_r1305874025
```suggestion
        const user = await this.userRepository.find({id});
```

File: backend/src/bundles/auth/auth.service.ts [hunk 10]
```diff
@@ -1,21 +1,60 @@
...
+        if (password !== 'HASH') {
+            throw new Error('Password not match');
+        }
+
+        return user.toObject();
+    }
+
+    public async login({ id }: { id: number }): Promise<UserSignInResponseDto> {
```
Comment (Monyster) [inline backend/src/bundles/auth/auth.service.ts]: Everywhere you use signIn, but here you use the login function
```suggestion
    public async signIn({ id }: { id: number }): Promise<UserSignInResponseDto> {
```

File: backend/src/bundles/auth/auth.controller.ts [hunk 11]
```diff
@@ -76,6 +93,51 @@ class AuthController extends ControllerBase {
...
+     */
+    private async signIn(
+        options: ApiHandlerOptions<{
+            body: UserSignInRequestDto;
+        }>,
+    ): Promise<ApiHandlerResponse> {
+        const user = await this.authService.verifyLoginCredentials(
+            options.body,
+        );
+        const login = await this.authService.login(user);
```
Comment (Monyster) [inline backend/src/bundles/auth/auth.controller.ts]:
```suggestion
        const signIn= await this.authService.signIn(user);
```

File: backend/src/bundles/auth/auth.controller.ts [hunk 12]
```diff
@@ -76,6 +93,51 @@ class AuthController extends ControllerBase {
...
+        return {
+            status: HttpCode.OK,
+            payload: login,
```
Comment (Monyster) [inline backend/src/bundles/auth/auth.controller.ts]:
```suggestion
            payload: signIn,
```

File: backend/src/bundles/auth/auth.service.ts (hunk 3 above)
Comment (Monyster) [reply backend/src/bundles/auth/auth.service.ts]: You can try to merge my branch #86, but my branch also uses JWT and this branch too, so it may take a long time to compare and fix.

cryptCompare already included in /backend/src/common/helpers

**Review by Monyster — CHANGES_REQUESTED** [review]

File: backend/src/bundles/auth/auth.ts (hunk 6 above)
Comment (JosuerBague) [reply backend/src/bundles/auth/auth.ts]: > The main reason I imported repository is because I wanted to make findByEmail method inside user-service return just user object if we will need to use it, while inside auth-service we need userEntity with password hash

We can still achieve this with just userService exposed to authService.

You can return a UserEntity via useRepo and then return user Object by running UserEntity.toObject and returning result from service.

See [userService.findByEmail](https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/88/files) sample flow.

**Comment by nikita-remeslov** [comment]
@ptarho merge development branch. We merged JWT to development

File: backend/src/bundles/users/user.entity.ts [hunk 13]
```diff
@@ -65,10 +65,12 @@ class UserEntity implements Entity {
     public toObject(): {
         id: number;
         email: string;
+        passwordHash: string;
```
Comment (Monyster) [inline backend/src/bundles/users/user.entity.ts:57]: Remove passwordHash.
Depending on the return statement https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/blob/d7721ebfa699a87a1ac4d8578dc1c5053b991efe/backend/src/bundles/users/user.service.ts#L42C14-L42C14, we can return passwords to the client in the browser. I think this is a bad idea.

Comment (ptarho) [reply backend/src/bundles/users/user.entity.ts:57]: I use findByEmail method inside auth-service to validate password. 
If I will remove it from return value I coludn`t use it.
We can switch usage of user-service into user-repository inside auth-service to fix that, but I am not sure about that

Comment (Monyster) [reply backend/src/bundles/users/user.entity.ts:57]: Oh, to compare the password, we need return it. Sorry, my mistake.

File: shared/tsconfig.json [hunk 14]
```diff
@@ -1,6 +1,6 @@
 {
     "extends": "../tsconfig.json",
-    "include": ["src"],
+    "include": ["src", "../backend/src/common/helpers/generate-salt"],
```
Comment (JosuerBague) [inline shared/tsconfig.json]: I think this might be causing issues?
Because when I remove this from tsconfig and run 'npm run build' inside shared it correctly builds the shared folder.

I might be wrong though.

File: backend/src/common/helpers/crypt-compare/crypt-compare.helper.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+import { compare } from 'bcrypt';
+
+const cryptCompare = async (
+    data: string,
+    encrypted: string,
+): Promise<boolean> => {
+    return await compare(data, encrypted);
```
Comment (MaksymGavatuk) [inline backend/src/common/helpers/crypt-compare/crypt-compare.helper.ts]: You don't need to _return await_ 
I have good article about it http://hassansin.github.io/Why-return-await-Is-a-Bad-Idea

File: backend/src/common/helpers/encrypt/encrypt.helper.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import { hash } from 'bcrypt';
+
+const encrypt = async (data: string, rounds: number): Promise<string> => {
+    return await hash(data, rounds);
```
Comment (MaksymGavatuk) [inline backend/src/common/helpers/encrypt/encrypt.helper.ts]: The same http://hassansin.github.io/Why-return-await-Is-a-Bad-Idea

File: backend/src/common/helpers/generate-salt/generate-salt.helper.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+import { genSalt } from 'bcrypt';
+
+const generateSalt = async (rounds: number): Promise<string> => {
+    return await genSalt(rounds);
```
Comment (MaksymGavatuk) [inline backend/src/common/helpers/generate-salt/generate-salt.helper.ts]: The same http://hassansin.github.io/Why-return-await-Is-a-Bad-Idea

File: backend/src/bundles/users/user.service.ts [hunk 18]
```diff
@@ -43,14 +45,25 @@ class UserService implements Service {
         };
     }
 
-    public async create(payload: UserSignUpRequestDto): Promise<UserEntity> {
-        return this.userRepository.create(
+    public async create(
+        payload: UserSignUpRequestDto,
+    ): Promise<UserFindResponseDto> {
+        const { PASSWORD_SALT_ROUNDS } = config.ENV.CRYPT;
```
Comment (MaksymGavatuk) [inline backend/src/bundles/users/user.service.ts]: save PASSWORD_SALT_ROUNDS as private field for UserService class and pass it in constructor instead of using from import

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: backend/src/bundles/users/user.service.ts (hunk 18 above)
Comment (ptarho) [reply backend/src/bundles/users/user.service.ts]: @MaksymGavatuk done

File: backend/src/bundles/auth/auth.service.ts [hunk 19]
```diff
@@ -1,33 +1,28 @@
...
-        const user = await this.userService.findById(id);
-
+    public async signIn(
+        userRequestDto: UserSignInRequestDto,
+    ): Promise<UserSignInResponseDto> {
+        const user = await this.verifyLoginCredentials(userRequestDto);
         return {
-            ...user,
-            token: await tokenService.create({ id }),
+            token: await tokenService.create({ id: user.id }),
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts:25]: why did you remove the user?

File: backend/src/bundles/auth/auth.service.ts [hunk 20]
```diff
@@ -46,40 +41,36 @@ class AuthService {
         }
 
         const user = await this.userService.create(userRequestDto);
-        const { id } = user;
-
         return {
-            ...user,
-            token: await tokenService.create({ id }),
+            token: await tokenService.create({ id: user.id }),
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts:45]: why did you remove the user?

File: backend/src/common/helpers/encrypt/encrypt.helper.ts
Comment (nikita-remeslov) [inline backend/src/common/helpers/encrypt/encrypt.helper.ts:1]: you dont need a helper. You already have token service

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

**Comment by nikita-remeslov** [comment]
@ptarho Pay attention to decode token logic - https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/88/files

File: backend/src/bundles/auth/auth.service.ts (hunk 19 above)
Comment (ptarho) [reply backend/src/bundles/auth/auth.service.ts:25]: https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/100#discussion_r1304781021

File: backend/src/common/helpers/encrypt/encrypt.helper.ts
Comment (ptarho) [reply backend/src/common/helpers/encrypt/encrypt.helper.ts:1]: This service is used for jwt token, but helpers is for bcrypt part
